internal/certdb/certdb.go

```diff
@@ -0,0 +1,136 @@
+// Package certdb provides a simplistic ORM to communicate with an SQL database for storage
+package certdb
+
+import (
+	"database/sql"
+	"fmt"
+
+	_ "github.com/mattn/go-sqlite3"
+)
+
+const queryCreateTable = "CREATE TABLE IF NOT EXISTS %s (CSR VARCHAR PRIMARY KEY UNIQUE NOT NULL, Certificate VARCHAR DEFAULT '')"
+
+const queryGetAllCSRs = "SELECT rowid, * FROM %s"
+const queryGetCSR = "SELECT rowid, * FROM %s WHERE CSR=?"
+const queryCreateCSR = "INSERT INTO %s (CSR) VALUES (?)"
+const queryUpdateCSR = "UPDATE %s SET Certificate=? WHERE CSR=?"
+const queryDeleteCSR = "DELETE FROM %s WHERE CSR=?"
+
+// CertificateRequestRepository is the object used to communicate with the established repository.
+type CertificateRequestsRepository struct {
+	table string
+	conn  *sql.DB
+}
+
+// A CertificateRequest struct represents an entry in the database.
+// The object contains a Certificate Request, its matching Certificate if any, and the row ID.
+type CertificateRequest struct {
+	ID          int
+	CSR         string
+	Certificate string
+}
+
+// RetrieveAll gets every CertificateRequest entry in the table.
+func (db *CertificateRequestsRepository) RetrieveAll() ([]CertificateRequest, error) {
+	rows, err := db.conn.Query(fmt.Sprintf(queryGetAllCSRs, db.table))
+	if err != nil {
+		return nil, err
+	}
+
+	var allCsrs []CertificateRequest
+	defer rows.Close()
+	for rows.Next() {
+		var csr CertificateRequest
+		if err := rows.Scan(&csr.ID, &csr.CSR, &csr.Certificate); err != nil {
+			return nil, err
+		}
+		allCsrs = append(allCsrs, csr)
+	}
+	return allCsrs, nil
+}
+
+// Retrieve gets a given CSR from the repository.
+// It returns the row id and matching certificate alongside the CSR in a CertificateRequest object.
+func (db *CertificateRequestsRepository) Retrieve(csr string) (CertificateRequest, error) {
+	var newCSR CertificateRequest
+	row := db.conn.QueryRow(fmt.Sprintf(queryGetCSR, db.table), csr)
+	if err := row.Scan(&newCSR.ID, &newCSR.CSR, &newCSR.Certificate); err != nil {
+		newCSR.ID = -1
+		newCSR.Certificate = ""
+		newCSR.CSR = ""
```

Comment:
I do not think this is necessary, the struct fields will be initialized to their zero values by default, so it will contain 0, "", "" by default. In Go, it is idiomatic to treat the returned value as invalid if an error is returned.

Reply:
That's nice. deleted it.